[DIR-741] e2e tests for events page

sebxian opened the pull request:
## Description

* Add e2e tests for the events page (events history and event listeners)

## Todo

* [x] Events history
* [x] Event listeners (waiting for input from Alan that will help me generate test data)
* [x] Update [wiki page](https://github.com/direktiv/wiki/wiki/Dev-(UI):-Writing-e2e-tests-with-playwright)

## Notes for Review

* I had to skip one "correct" test. Pagination of event listeners is broken on the test server, but works on main. Thus I have created this ticket: https://linear.app/direktiv/issue/DIR-930/update-e2e-tests-and-gitlab-runner-to-current-main. Let me know if you agree with doing this soon.
* I made an effort to apply best practices while writing the test, based on the experience gathered over the last months, and wrote a [wiki page](https://github.com/direktiv/wiki/wiki/Dev-(UI):-Writing-e2e-tests-with-playwright) on that. Please provide feedback on the wiki page along with this review if possible (the wiki page isn't completely finished, I think this is an ongoing project).
* I also refactored the way locators are used in the test for the Instances page, because the way that testids were used here caused problems for the tests on the events page. See the wiki page for the general thoughts behind this.
* A stylistic choice we could discuss: I added some comments to delineate sections in the longer tests, using this format:
  ```js  
  /* filter by status "complete", expect 2 results to be rendered */
  await btnPlus.click();
  await page.getByRole("option", { name: "status" }).click();
  await page.getByRole("option", { name: "complete" }).click();
  ```
  instead of double slashes (`// comment`).
  I think they look a bit nicer, but ideally we'll be consistent about that.
  In one of my old teams we decided to use `/* comment */` for documentation and `// Todo: fix this` for temporary notes that will eventually be deleted and I quite liked that pattern.
* my test for event listeners in `e2e/events/listeners.spec.ts` is comparably short. I thought about adding more detail but decided against it. Instead, we should have a discussion about a separation of concerns on different levels of testing. 

### Thoughts on the scope of frontend e2e tests

I would argue that:
 * e2e tests should focus on the main requirements of the implementation, and on the interplay between api and ui. But neither api behavior nor ui functions should be tested in all detail for every single page.
* In this case I made sure all relevant information is visible. Navigation to the workflow-detail route works.
* I think it is a bit too much to to test tooltips or copy-to-clipboard functions again on every single page. It should be enough to have an integration test on another level of testing that makes sure the component works in principle (we don't have that layer of testing yet).
* I made an exception for pagination. It's a bit annoying to always create a ton of test data to test pagination. But the implementation is not identical across different pages, so we probably cannot avoid testing every single use case.
* The full range of backend behavior and possible display values needs to be tested in api response tests. For example, the only event listener type I get in this test is "start workflow". Other types are possible, but not all of them can easily be "simulated" in an e2e test. So I think the main point of the test is to assure that the frontend renders a text string received from the api, but not to test every possible use case.
 
What do you think? It would be valid to argue against leaving these gaps in the tests, and if we decide this is ok, we may need to define a multi-layer testing strategy that addresses the different layers of integration on which our app or single parts of it may fail.

stefan-kracht reviewed (changes requested):
Very good work

> I had to skip one "correct" test. Pagination of event listeners is broken on the test server, but works on main. Thus I have created this ticket: https://linear.app/direktiv/issue/DIR-930/update-e2e-tests-and-gitlab-runner-to-current-main. Let me know if you agree with doing this soon.

No problem. I just updated from the base branch and some [other tests](https://github.com/direktiv/direktiv-ui/actions/runs/6609251949/job/17949717793?pr=506) are also broken. This is probably the same issue and we can skip them as well?

I like what you put in the wiki a lot, very good points and also very good explanations to all of them. I also agree on what you wrote about the scope of the e2e tests. I fixed a [small typo](https://github.com/direktiv/wiki/wiki/Dev-(UI):-Writing-e2e-tests-with-playwright/_compare/66413ccfacf811c64941f759e79b078a6c6197cf) the other day. 

I would like to propose a section about not adding additional DOM elements just for testing (e.g. adding a wrapper div to have a better selector can cause potential side effects like breaking CSS selectors or breaking flex box which heavily relies on how its children are composed)

**About comments:** 
I am very open to a general agreement on how we comment, but my chain of thoughts is: when we want to have a rule about comments, it should be an eslint rule (every developer naturally has their own habit on how to comment) and I would need one to get use to /* */ for one line comments (but I would certainly do that) because it is a different shortcut. 

I noticed `e2e/events/history.spec.ts` has some one-line-comments that go over three lines, and I'm not sure if this is correct according to your comment suggestions or is there an additional patter/rule for that?

sebxian commented on e2e/events/listeners.spec.ts:
```diff
@@ -0,0 +1,139 @@
+    'type "start workflow" is rendered for every event listener'
+  ).toHaveCount(4);
+
+  await workflowNames.map((name, index) =>
```
Good catch, fixed in fc15e5f9727d87710dff45ea0510057903edf1d9

sebxian commented:
Thanks for the review.

> I just updated from the base branch and some [other tests](https://github.com/direktiv/direktiv-ui/actions/runs/6609251949/job/17949717793?pr=506) are also broken. This is probably the same issue and we can skip them as well?

I am investigating them now.
 
> I would like to propose a section about not adding additional DOM elements just for testing (e.g. adding a wrapper div to have a better selector can cause potential side effects like breaking CSS selectors or breaking flex box which heavily relies on how its children are composed)

Good point. Added here: https://github.com/direktiv/wiki/wiki/Dev-(UI):-Writing-e2e-tests-with-playwright#dont-add-dom-elements-just-for-testing

> **About comments:** I am very open to a general agreement on how we comment, but my chain of thoughts is: when we want to have a rule about comments, it should be an eslint rule (every developer naturally has their own habit on how to comment) and I would need one to get use to /* */ for one line comments (but I would certainly do that) because it is a different shortcut.

I think a linting rule would overcomplicate things in this case. My idea is to have both `/* structure / documentation */` comments and `// temporary comments` (and the idea only applies to test files). So it's a case by case decision and ESLint will not be able to tell the difference.

> I noticed `e2e/events/history.spec.ts` has some one-line-comments that go over three lines, and I'm not sure if this is correct according to your comment suggestions or is there an additional patter/rule for that?

In this case some of the comments don't fit in one line so I went with the same pattern with most of them. It's a bit inconsistent but I don't think we need to be 100% strict on this.

I wouldn't want to overcomplicate the rules for comments by making strict definitions. But if you agree I'll use `/* this style */` comments for the use case described and see if that pattern catches on.

stefan-kracht reviewed (approved):
Nice. Thanks for adding the paragraph to the wiki. 

I will try to use this comment pattern for e2e tests as well and let's see how it goes
